README
[You should read IMPORTANT_NOTES which notes version-specific issues]


VRPipe is a generic pipeline system designed for use by the Vertebrate
Resequencing team at the Sanger Institute, but also for broad use for anyone.

For an overview of why it was created, read our vision document:
https://github.com/VertebrateResequencing/vr-pipe/wiki/Vision

It comprises mostly self-documented Perl code. There are both scripts and 
modules within subfolders.

Help on each module is currently of limited availability, but eventually all
will be self-documented with POD, which can be read with eg. perldoc.

Each script has a --help option, eg:
vrpipe-status --help



INSTALLATION
------------

For use on Amazon's cloud, see:
https://github.com/VertebrateResequencing/vr-pipe/wiki/AWS

You will need the source version of samtools compiled with -fPIC and -m64 in the
CFLAGS, and the environment variable SAMTOOLS pointing to that source directory
(which should now contain bam.h and libbam.a, and MUST contain the samtools
executable). Samtools can be downloaded from here:
http://sourceforge.net/projects/samtools/files/samtools/
(Note that tests use the samtools in your PATH, not the samtools in your
SAMTOOLS dir; production pipelines will use the samtools at the absolute path
you configure them with during setup - the default will be the first one in your
PATH.)

You need Redis 2.6 or greater installed. Compile it as normal ('make') and
simply include redis-server in your PATH. You do not have to configure it in any
way. But note that its maximum number of connections is capped by the max file
descriptors limit imposed on the user that will run vrpipe-server, so make sure
that limit (ulimit -n) is higher than the number of CPUs in your cluster.
Redis can be downloaded from here: http://redis.io/download

It is recommended that you set PERL_INLINE_DIRECTORY to ~/.Inline and create
that directory.

You require Module::Build in order for the Build.PL script to work. It is
recommended you 'install Bundle::CPAN' using 'cpan' prior to attempting setup.


First, do the initial setup and dependency installation:
$ perl Build.PL

If this says you have "ERRORS/WARNINGS FOUND IN PREREQUISITES" try:
$ ./Build installdeps
to install missing prerequisites from CPAN.

installdeps will install things automatically in some random order, but ideally
you should manually install certain modules in a specific order:
1) EV
2) AnyEvent
3) Continuity (specifically its Coro dependency)

Possible problems you may encounter in installing your missing prerequisites:
# You may find that you have to unset your LANG environment variable temporarily
  to install all CPAN prerequisites.
# You will likely find that Inline::Filters fails its preprocess.t test; this is
  ok and you can manually use 'cpan' to:
  cpan> force install Inline::Filters
# You may find that MooseX::Types::Parameterizable fails its tests. In that case
  you can manually use 'cpan' to install an earlier version that should work:
  cpan> install JJNAPIORK/MooseX-Types-Parameterizable-0.07.tar.gz
# If you find that AnyEvent::ForkManager fails tests and emits errors
  mentioning WNOHANG, you will have to cd to the cpan build directory and edit
  line 11 of blib/lib/AnyEvent/ForkManager.pm to read "use POSIX qw(WNOHANG);"
  (without the quotes). Then you can do "make test" and "make install".
# Twiggy may fail its disconnect.t test. This is safe to ignore; just:
  cpan> force install Twiggy
# You may encounter difficulties installing Proc::ProcessTable, which is one of
  Proc::Killfam's dependencies. In that case, try manually installing an earlier
  version of it before retrying the Proc::Killfam install normally:
  cpan> install DURIST/Proc-ProcessTable-0.44.tar.gz
# Some installs of Perl 5.8.8 can have trouble with Class::Accessor::Grouped (a
  prerequisite of DBIx::Class), where some tests fail due to desired error
  messages not looking quite right. This should be harmless in reality, so just
  force install it:
  cpan> force install Class::Accessor::Grouped
  and then try the DBIx::Class installation again normally.
# If you run into a problem installing XML::LibXML due to an inability to find
  the required libs, try using your OS package management system to install it,
  eg. for Ubuntu:
  sudo apt-get install libxml-libxml-perl
# Likewise for XML::LibXSLT:
  sudo apt-get install libxml-libxslt-perl
# Crypt::Random has a prerequisite of Math::Pari, which in turn requires pari-gp
  and may have trouble finding and using it. If you encounter problems, try
  entering the cpan build directory where it tried to build Math::Pari, then
  do:
  sudo wget ftp://megrez.math.u-bordeaux.fr/pub/pari/unix/OLD/pari-2.3.5.tar.gz
  sudo perl Makefile.PL Configure pari_tgz=./pari-2.3.5.tgz
  sudo make; sudo make test; sudo make install
# Time::Format may fail some tests with errors mentioning problems with
  Date::Manip. As long as you actually have Date::Manip installed, these
  failures are probably fine and you can force the install:
  cpan> force install Time::Format
# Redis may fail one its tests due to an error message not being in the
  expected format. This is safe to ignore:
  cpan> force install Redis
# If all the prerequisites seem to be installed but running 'perl Build.PL'
  gives you an error message mentioning problems with any kind of Moose module,
  one possibility is that you have a mixture of some new Moose modules and old
  ones. You can solve this by updating all your Moose-related modules to their
  latest versions. Eg. if you have cpanm installed, try:
  $ cpanm Class::MOP Fey Fey::ORM MooseX::Aliases MooseX::ClassAttribute
    MooseX::LazyRequire MooseX::Method::Signatures MooseX::NonMoose MooseX::POE
    MooseX::Role::Cmd MooseX::Role::Parameterized MooseX::SemiAffordanceAccessor
    MooseX::Singleton MooseX::StrictConstructor
# If you get as far as running "./Build test" (see below), and find some tests
  failing with an error message that mentions IO::Uncompress::Gunzip, try
  installing the latest version of that:
  cpan> install IO::Uncompress::Gunzip

Once the prerequisites are installed you will be asked a number of setup
questions such as the details of your production and testing databases. It is
recommended to use a stand-alone relational database such as MySQL. If you're
only using code such as a parser or the bas function, or only running pipelines
on very small datasets, you can use SQLite, supplying a file path for the
database name. However SQLite is not recommended for production use.

Note that when using MySQL, the server should be configured to allow at least
1/3 as many connections as you have CPUs in your cluster. It must also allow
the transaction isolation level to be changed from a default of REPEATABLE-READ
to READ-COMMITTED, and gap-locks must be possible
(innodb_locks_unsafe_for_binlog must be off). In practice this means that if
you have database replication set up, you may find that you have to reconfigure
the server to binlog-mode=MIXED and restart/redo your replication server.
To avoid possible issues we recommend MySQL 5.5 with the following settings in
my.cnf, where the idea is to maximise transaction concurrency and performance:
max_connections = 1000
binlog_cache_size=32M
innodb_log_file_size=512M
innodb_log_files_in_group = 2
innodb_file_per_table = 1
innodb_fast_shutdown=0
innodb_buffer_pool_size = 1G
innodb_change_buffering=all
innodb_flush_log_at_trx_commit=0
innodb_log_buffer_size = 32M
innodb_thread_concurrency=16
innodb_concurrency_tickets=5000
innodb_commit_concurrency=50
innodb_autoinc_lock_mode=2
binlog_format=mixed
transaction_isolation = REPEATABLE-READ
innodb_locks_unsafe_for_binlog = 0

The SGE (Sun|Oracle GridEngine) scheduler has only been tested with SGE 8.1.3
from: https://arc.liv.ac.uk/trac/SGE. If tests get stuck and nothing happens
with no obvious error messages, try running:
$ perl -Imodules -It scripts/vrpipe-server --deployment testing --farm testing_farm -f --debug start
which should provide an error message. You may then have to alter or subclass
VRPipe::Schedulers::sge; bug reports or patches welcome.


Once setup is complete modules/VRPipe/SiteConfig.pm will have been created. It
is this file that holds your site-wide configuration of VRPipe.


To test the code prior to using it:
$ ./Build test

(Note that there is currently an issue where the first time you ever run tests
they will fail because the test database wasn't created yet; start the tests
running until you get at least one 'ok' message, then ctrl-c to kill the tests,
then start them again: they should now work)

A number of environment variables affect which tests run and how. In most cases
it is fine to not worry about these extra variables. They are described here for
the sake of completeness.
VRPIPE_TEST_PIPELINES, when true, will fully test all the pipelines and steps
                       instead of just the core functionality of the system.
                       These tests take a very long time - potentially hours -
                       so this is off by default. If you're interested in
                       running a particular pipeline, you can test just that
                       pipeline by setting this variable to true and then
                       running:
               $ ./Build test --test_files t/VRPipe/Pipelines/[name].t --verbose
GATK, pointing to a directory containing GATK jar files, will enable tests that
      require GATK
GATK2, pointing to a directory containing GATK version 2+ jar files, will enable
      tests that require GATK version 2 or higher
PICARD, pointing to a directory containing Picard jar files, will enable tests
      that require Picard
CRAMTOOLS, pointing to a directory containing Cramtools jar files, will enable
      tests that require Cramtools
TRIMMOMATIC_JAR_PATH, pointing to the timmomatic .jar file, will enable tests
      that require Trimmomatic
CONVEX_R_LIB, pointing to a directory containing your R lib files, will enable
      tests of Convex
VRPIPE_VRTRACK_TESTDB will enable testing of the VRTrack DataSource, using the
                      value supplied as the database name (other database
                      details will come from the standard VRTrack environment
                      variables)
Additionally, tests that require certain executables will not run unless you
have those executables in your PATH.


To install:
$ ./Build install
(or just include the modules subdirectory in your PERL5LIB, and include the
scripts subdirectory in your PATH)

To create your production database (testing database is created automatically):
$ vrpipe-db_deploy

If in the future the VRPipe code is updated and there is a change to the schema,
you will need to stop VRPipe, update your code, then run:
$ vrpipe-db_upgrade


To use:
Ensure that vrpipe-server is always running on a machine that can submit jobs
to your job scheduler (it must also be able to ssh to all other machines without
a password), and that has at least 2GB of free memory (which probably means the
machine must have at least 3GB physical memory), especially if using the local
scheduler:
$ vrpipe-server --farm my_farm start
Now you can use vrpipe-create_step and vrpipe-create_pipeline to create
pipelines if necessary. Then end-users can set up a pipeline using vrpipe-setup.
vrpipe-setup and vrpipe-status will be the most commonly used scripts. The web
front-end is currently limited to providing status, but it does so a bit nicer
and a bit faster than vrpipe-status cmd line tool. You can alter what it
displays by manually appending vrpipe-status arguments to the url, eg.
http://[serverurl]/status?brief=1&incomplete=1&user=me

Access to output files can be achieved with vrpipe-output. When you have an
output file but can't remember how it was made, use vrpipe-fileinfo.

As a general point, any Perl script you write that wants to use some VRPipe
code should typically "use VRPipe;". After that most things should work without
further "use" statements. To choose the deployment you can "import" the one you
want. eg. to work with your testing database:
    use VRPipe 'testing';
or on the command-line:
$ perl -MVRPipe=testing -e '...'


When things go wrong:
For normal errors caused by failing jobs in a pipeline, and for most other
issues, you should receive an email that either specifies the problem, or asks
you to investigate (and fix) with vrpipe-submissions. Using vrpipe-status will
also point out issues in case you're not receiving emails.

In the rare event that vrpipe-status and vrpipe-submissions do not reveal the
problem, you can investigate further by looking at the logs. vrpipe-server keeps
a log file in the logging directory you configured during installation (when you
ran 'perl Build.PL'). For problems with a particular setup, use vrpipe-logs to
see what has been happening with it.

Sometimes it can happen that a step is failing because the output of a previous
step is corrupt. VRPipe will only ever automatically keeps retrying the current
step, so it will keep failing. vrpipe-submissions only gives you an easy way to
force the reattempt of a failed submission (ie. which will be for the current
step), so it isn't helpful either. In this case you should use vrpipe-elements,
which provides an easy way to 'go back' and reset steps that VRPipe thought
completed successfully. Just be careful not to redo any step that outputs files
that might be used by other running jobs in your setup, or by other setups.



EXTERNAL SOFTWARE
-----------------

Some software need environment variables setup (using setenv in csh or export in
bash). The following list shows the name of the software, the environment
variable you need to set, and the value you should set it to, separated by
commas.

samtools,SAMTOOLS,/path/to/samtools/source_directory
cramtools,CRAMTOOLS,/path/to/cramtools_jar_files

eg. to have cramtools work properly in the pipelines you might do:
setenv CRAMTOOLS /path/to/cramtools_jar_files
or
export CRAMTOOLS=/path/to/cramtools_jar_files
depending on what shell you are using.

These values are used internally and are independent of values you might set
when asked for eg. the path to your cramtools jars when running vrpipe-setup.

A number of other environment variables are used to setup default values for
when running vrpipe-setup, but are not required to use the software (ie. you can
just manually enter the required path each time you setup the pipeline):

GATK,GATK,/path/to/GATK_jar_files
GATK2,GATK2,/path/to/GATK2_jar_files
picard,PICARD,/path/to/picard_jar_files
R,R_LIB,/path/to/R_library_files
bismark,BISMARK_GENOME_FOLDER,/path/to/bismark_genome_files
trimmomatic,TRIMMOMATIC_JAR_PATH,/path/to/trimmomatic_jar_file


COPYRIGHT & LICENSE
-------------------
Copyright (c) 2011-2013 Genome Research Limited.

This file is part of VRPipe.

VRPipe is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see L<http://www.gnu.org/licenses/>.

The usage of a range of years within a copyright statement contained within this
distribution should be interpreted as being equivalent to a list of years
including the first and last year specified and all consecutive years between
them. For example, a copyright statement that reads 'Copyright (c) 2005, 2007-
2009, 2011-2012' should be interpreted as being identical to a statement that
reads 'Copyright (c) 2005, 2007, 2008, 2009, 2011, 2012' and a copyright
statement that reads "Copyright (c) 2005-2012' should be interpreted as being
identical to a statement that reads 'Copyright (c) 2005, 2006, 2007, 2008, 2009,
2010, 2011, 2012'."

HISTORY
0.213   Critical locking overhaul resolves vrpipe-server not functioning
        correctly.
        Fixed vrpipe-output so that it works again.
        Fixed all vrpipe-* scripts and processes to not spawn extraneous redis
        servers; now only vrpipe-server will spawn 1 as needed.
        bcftools_cnv step fixed to get true sample name from vcf.
        bcf_to_vcf step improved so that $bcftools can be replaced by
        bcftools_exe in post_calling_vcftools option.
        bam_import_from_irods_and_index step fixed so that it can be installed.
0.212   Critical speed fix for steps that take datasource input with multiple
        types and many files.
        cram_index and bam_index steps now die if they can't delete an out-of-
        date index file.
        Overhauled internal locking mechanism should result in a slightly faster
        system with fewer, shorter-lived stalls.
        (Sanger specific:) irods warehouse datasource and npg_cram_stats_parser
        step now default to getting F0xB00 stats files instead of F0x900. Also,
        the correct studies according to warehouse are related to the input
        file, ignoring study information stored in irods metadata. irods
        datasource no longer falls over if given a query file with blank lines.
        GATK steps can now use a gatk_key to turn off phoning home, enabled by
        default if environment variable GATK_KEY is set.
        New gatk_genotype_gvcfs pipeline which can do the calling to chrom vcf
        in one setup (using both datasource and step chunking).
        New pipelines: bam_calling_with_gatk_haplotype_caller,
        bam_calling_with_mpileup_via_bcf, bam_import_from_irods_and_index,
        bam_indel_realignment_and_bqsr, bam_merge_and_mark_duplicates,
        bam_merge_and_streaming_mark_duplicates,
        gvcf_calling_one_combine_with_gatk_haplotype_caller,
        gvcf_calling_two_combine_with_gatk_haplotype_caller,
        gvcf_calling_with_gatk_haplotype_caller,
        sample_bam_remapping_with_bwa_mem
        Deleted pipelines (some of them renamed to new ones above):
        sample_gvcf_via_bcf_with_mpileup,
        sample_gvcf_with_gatk_haplotype_caller,
        sample_indel_realignment_and_bqsr,
        sample_irods_import_and_mark_duplicates,
        sample_irods_import_and_streaming_mark_duplicates
0.211   bam_improvement_gatk_v2 pipeline fixed to work again.
        irods_sample_import pipeline renamed as
        sample_irods_import_and_mark_duplicates.
        bam_reheader step fixed to work when reads metadata on input file is
        wrong.
        bcftools_cnv step fixed to always remove invalid plot<-sample
        relationships.
0.210   Various bam-outputting steps are now more tolerant of input file
        reads metadata being wrong.
        bcftools_cnv step fixed to properly relate all plots to the sample node.
        New sample_indel_realignment_and_bqsr pipeline.
        Critical speed fix for vrtrack_update_mapstats step.
        Fixed polysomy step so the output files have the correct sample name
        metadata.
0.209   Minor fix for vcf_merge_different_samples_control_aware step when the
        samples have no control.
        Fix for the polysomy step when the vcf has duplicate samples.
        Fixed step post processing when an output file needs permission changes
        and wasn't output by anything.
0.208   Critical speed fix for graph_filter queries in datasources (allowing
        those datasources to work now).
        Further critical speed fixes for processing all pipelines, allowing them
        to proceed.
0.207   Critical speed fix for processing all pipelines (especially evident when
        unstalling a stalled/started-from-scratch setup)
        vrpipe-setup fixed to allow creating new setups again.
        bwa_mem_to_bam fixed to solve conflict issues with samtools sort.
        Various steps in the polysomy_cnv_caller pipeline have been fixed to
        work as intended.
0.206   Proper fix for v0.205.
0.205   Fix for hipsci QC to handle _CTRL appended to sample names in files.
0.204   CNV and Pluritest-related steps now store their results properly in the
        graph database against sample and donor nodes.
        vrpipe-setup now allows reactivate/deactivate of multiple setups.
        New sample_gvcf_via_bcf_with_mpileup pipeline.
        New sample_gvcf_with_gatk_haplotype_caller pipeline.
        vrtrack qc website donor view now allows setting donor sample qc status.
0.203   Fixed vrpipe-setup to handle new default options in the irods
        datasource correctly.
0.202   npg_cram_stats_parser step now stores md5_of_ref_seq_md5s
0.201   irods datasource methods fixed to all work with the new irods protocol
        system.
0.200   delete_inputs step behaviour now checks that input files are not needed
        by another setup before deleting.
        New pipeline for calling with fermikit.
        Bwa mem steps can now take sequence dict file as input.
        Steps can get chunks across the gnome with the new
        StepGenomeChunkingRole.
        npg_cram_stats_parser step now compares cram header to expectations,
        with result stored in new Header_Mistakes graph node.
0.199   This is a big change from the previous version. Re-read README.md for
        installation changes and study IMPORTANT_NOTES if upgrading.
        Newer steps and pipelines, and some vrpipe internal and website
        features, use the Neo4J graph database to store information. Neo4J may
        eventually replace the use of MySQL/postrgres.
        Issues with running out of database connections have been (mostly)
        resolved, so you can increase or get rid of --max_submissions arg to
        vrpipe-server.
        Now multiple setup handlers are run on the farm, each handling a few
        setups, to spread the load and help avoid some setups never being
        updated because a single setup handler never got around to dealing with

share/js/vrpipe.js
// we have a custom ko binding to handle our loading feedback
// (from http://blog.greatrexpectations.com/2012/06/17/loading-placeholders-using-knockout-js/)
ko.bindingHandlers.loadingWhen = {
    init: function (element) {
        var $element = $(element),
            currentPosition = $element.css("position")
            $loader = $("<div>").addClass("loader").hide();
        
        //add the loader
        $element.append($loader);
        
        //make sure that we can absolutely position the loader against the original element
        if (currentPosition == "auto" || currentPosition == "static")
            $element.css("position", "relative");

        //center the loader
        $loader.css({
            position: "absolute",
            top: "20px",
            left: "50%",
            "margin-left": -($loader.width() / 2) + "px"
        });
    },
    update: function (element, valueAccessor) {
        var isLoading = ko.utils.unwrapObservable(valueAccessor()),
            $element = $(element),
            $childrenToHide = $element.children(":not(div.loader)"),
            $loader = $element.find("div.loader");

        if (isLoading) {
            $childrenToHide.css("visibility", "hidden").attr("disabled", "disabled");
            $loader.fadeIn("slow");
        }
        else {
            $loader.fadeOut("fast");
            $childrenToHide.css("visibility", "visible").removeAttr("disabled");
        }
    }
};

// we can have sortable table columns
// (from http://stackoverflow.com/a/16964843/675083)
ko.bindingHandlers.sort = {
    init: function(element, valueAccessor, allBindingsAccessor, viewModel, bindingContext) {
        var asc = false;
        element.style.cursor = 'pointer';
        
        element.onclick = function(){
            var value = valueAccessor();
            var prop = value.prop;
            var data = value.arr;
            
            asc = !asc;
            
            data.sort(function(left, right){
                var rec1 = left;
                var rec2 = right;
                
                if(!asc) {
                    rec1 = right;
                    rec2 = left;
                }
                
                var props = prop.split('.');
                for(var i in props){
                    var propName = props[i];
                    var parenIndex = propName.indexOf('()');
                    if(parenIndex > 0){
                        propName = propName.substring(0, parenIndex);
                        rec1 = rec1[propName]();
                        rec2 = rec2[propName]();
                    } else {
                        rec1 = rec1[propName];
                        rec2 = rec2[propName];
                    }
                }
                
                return rec1 == rec2 ? 0 : rec1 < rec2 ? -1 : 1;
            });
        };
    }
};

// load knockout templates from an external file
// from http://stackoverflow.com/questions/9480697/using-an-external-template-in-knockoutjs/22902067#22902067
// use like:
// <div id="nav">
//     <script type="text/html" src="/login_template.html" id="login-template"></script>
//     <div data-bind="template: { name: 'login-template' }"></div>
// </div>
// var livm = new ko.LogInViewModel();
// loadExternalKnockoutTemplates(function() {
//     ko.applyBindings(livm, $('#nav')[0]);
// });
function loadExternalKnockoutTemplates(callback) {
    var sel = 'script[src][type="text/html"]:not([loaded])';
    $toload = $(sel);
    function oncomplete() {
        this.attr('loaded', true);
        var $not_loaded = $(sel);
        if(!$not_loaded.length) {
            callback();
        }
    }
    ko.utils.arrayForEach($toload, function(elem) {
        var $elem = $(elem);
        $elem.load($elem.attr('src'), oncomplete.bind($elem));

    });
}

// a knockout view model for logging in, which can be 'inherited' in another
// viewmodel (for eg. your navbar) by doing:
// ko.LogInViewModel.call(self);
// Or to just use it directly:
// var livm = new ko.LogInViewModel(); ko.applyBindings(livm, $('#nav')[0]);
// If the user successfully authenticates, the username() observable will
// contain their username.
// You must have html elements with ids loginUsername, loginPassword,
// loginSubmit and loginDropdown on your page for this to work. You should
// show the loginError() observable somewhere as well. Best to just use the
// login_template.html (see comments for loadExternalKnockoutTemplates).
(function (ko, undefined) {
    ko.LogInViewModel = function () {
        var self = this;
        self.loading = ko.observableArray();
        self.errors = ko.observableArray();
        self.username = ko.observable();
        self.loginError = ko.observable();
        
        self.getUserName = function() {
            // ajax call to refresh the idle session timeout and populate
            // username if user has already logged in
            vrpipeRestMethod('authenticated_user', 'n/a', {}, self.loading, self.errors, function(data) {
                self.username(data['user']);
            });
        }
        
        // to set username immediately if logged in, and to keep the session
        // alive if we are already logged in, first we call getUserName() on
        // page load, then we call it every hour if there has been any activity
        // on the page in the last hour
        self.getUserName();
        self.intervalMilliSeconds = 3600000;
        self.hadActivity = false;
        $(document.body).bind("mousemove keypress", function(e) {
            self.hadActivity = true;
        });
        self.maintainSession = function() {
            if (self.hadActivity) {
                self.getUserName();
            }
            self.hadActivity = false;
        }
        setInterval(self.maintainSession, self.intervalMilliSeconds);
        
        self.login = function() {
            var user = $("#loginUsername").val();
            var password = $("#loginPassword").val();
            if (user && password) {
                $('#loginSubmit').prop('disabled', true);
                vrpipeRestMethod('authenticate', 'n/a', { user: user, password: password }, self.loading, self.errors, function(data) {
                    $('#loginSubmit').prop('disabled', false);
                    if (data['user']) {
                        $('#loginDropdown').dropdown('toggle');
                        self.username(data['user']);
                        window.location.reload();
                    }
                    else {
                        self.loginError('Failed; try again?');
                    }
                });
            }
        }
        
        self.logout = function() {
            vrpipeRestMethod('deauthenticate_user', 'n/a', { }, self.loading, self.errors, function(data) {
                self.loginError(undefined);
                self.username(undefined);
                window.location.reload();
            });
        }
    };
}(ko));

// ajax function to send/get json queries
var ajax = function(uri, method, loading, error, data) {
    var loadingTimeout;
    var request = {
        url: uri,
        type: method,
        contentType: "application/json",
        accepts: "application/json",
        cache: false,
        dataType: 'json',
        data: JSON.stringify(data),
        beforeSend: function (xhr) {
            // say that we're loading, but only if for more than
            // half a second
            if (loadingTimeout) { clearTimeout(loadingTimeout); }
            loadingTimeout = setTimeout(function() {
                loading.push(true);
            }, 500);
        },
        error: function(jqXHR, textStatus, errorThrown) {
            if (textStatus) {
                var msg = "Ajax " + textStatus + " " + jqXHR.status + " connecting to " + uri;
                if (errorThrown) {
                    msg = msg + ": " + errorThrown;
                }
                error.push(msg);
            }
        },
        complete: function(jqXHR, textStatus) {
            if (loadingTimeout) { clearTimeout(loadingTimeout); }
            loading.pop();
        }
    };
    return $.ajax(request);
}

// function to call a VRPipe Rest method and do something with the
// json response, and its helper function to deal with errors
var vrpipeRestDataErrorParser = function(data, error) {
    if (data == null) {
        error.push('no data');
        return 1;
    }
    if (data.hasOwnProperty('errors')) {
        for (var i = 0; i < data.errors.length; i++) {
            error.push(data.errors[i]);
        }
        return 1;
    }
    return 0;
}
var vrpipeRestMethod = function(category, method, args, loading, error, donefunction, other) {
    var promise = ajax('/rest/' + category + '/' + method, 'POST', loading, error, args);
    
    if (donefunction) {
        promise.done(function(data) {
            if (! vrpipeRestDataErrorParser(data, error)) {
                donefunction(data, other);
            }
        });
    }
    else {
        return promise;
    }
}

var shortenString = function(string) {
    var shortened = string;
    if (string.length > 25) {
        shortened = string.substring(0, 5) + '[...]' + string.slice(-15);
    }
    return shortened;
}

var wbr = function(string) {
    if (! string) { return '' }
    var wbrd = '' + string;
    wbrd = wbrd.replace(/([_\-\W])/g, "$1<wbr>");
    wbrd = wbrd.replace(/( <wbr>)/g, " ");
    return wbrd;
}

var zeropad = function(num) {
    return (num < 10) ? ("0" + num) : num;
}

// convert epoch seconds to friendly date
var epochToDate = function(epoch) {
    var date = new Date(epoch * 1000);
    return date.getFullYear() + "/" + zeropad(date.getMonth() + 1) + "/" + zeropad(date.getDate()) + " " + zeropad(date.getHours()) + ":" + zeropad(date.getMinutes()) + ":" + zeropad(date.getSeconds());
}

// find out if 2 arrays are equal, regardless of order
function arraysEqual(a, b) {
    if (a === b) return true;
    if (a == null || b == null) return false;
    if (a.length != b.length) return false;

    a.sort();
    b.sort();

    for (var i = 0; i < a.length; ++i) {
        if (a[i] !== b[i]) return false;
    }
    return true;
}

// split a comma separated string of numbers in to an array of ints
function splitToInts(str) {
    return $.map(str.split(','), function(value){return parseInt(value, 10);});
}

// wait until a given observeable array has length, then run the given function
var runWhenPopulated = function(observable, runFunction) {
    if (observable().length) {
        runFunction();
    }
    else {
        var subscription;
        subscription = observable.subscribe(function(newValue) {
            if (newValue.length) {
                subscription.dispose();
                runFunction();
            }
        });
    }
}

// get a stacktrace
var stackTrace = function() {
    var err = new Error();
    console.log(err.stack);
}